File: backend/app/models/trip.py
import uuid
from datetime import datetime

from sqlalchemy import Boolean, CheckConstraint, DateTime, ForeignKey, Index, Integer, Numeric, String
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base
from app.models.base import TimestampMixin, UUIDPKMixin

TRIP_STATUSES = ("planificado", "en_curso", "completado", "cancelado")


class Trip(Base, UUIDPKMixin, TimestampMixin):
    __tablename__ = "trips"
    __table_args__ = (
        CheckConstraint(f"status IN {TRIP_STATUSES}", name="ck_trips_status"),
        Index("ix_trips_company_status", "company_id", "status"),
        Index("ix_trips_vehicle_id", "vehicle_id"),
        Index("ix_trips_driver_id", "driver_id"),
    )

    company_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("companies.id", ondelete="CASCADE"), nullable=False
    )
    vehicle_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("vehicles.id"), nullable=False
    )
    driver_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("drivers.id"), nullable=False)
    trailer_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("vehicles.id"), nullable=True
    )
    origin: Mapped[str] = mapped_column(String(255), nullable=False)
    destination: Mapped[str] = mapped_column(String(255), nullable=False)
    distance_km: Mapped[float | None] = mapped_column(Numeric(10, 2), nullable=True)
    cargo_type: Mapped[str] = mapped_column(String(100), nullable=False)
    is_round_trip: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="planificado")
    start_odometer_km: Mapped[int | None] = mapped_column(Integer, nullable=True)
    end_odometer_km: Mapped[int | None] = mapped_column(Integer, nullable=True)
    freight_cost: Mapped[float | None] = mapped_column(Numeric(10, 2), nullable=True)
    advance_payment: Mapped[float] = mapped_column(Numeric(10, 2), nullable=False, default=0)
    started_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    ended_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    vehicle: Mapped["Vehicle"] = relationship(foreign_keys=[vehicle_id])
    trailer: Mapped["Vehicle | None"] = relationship(foreign_keys=[trailer_id])
    driver: Mapped["Driver"] = relationship()
    expenses: Mapped[list["TripExpense"]] = relationship(
        back_populates="trip", cascade="all, delete-orphan"
    )
    payroll: Mapped["TripPayroll | None"] = relationship(
        back_populates="trip", cascade="all, delete-orphan", uselist=False
    )

File: frontend/src/types/trip.ts
export type TripStatus = 'planificado' | 'en_curso' | 'completado' | 'cancelado'

export interface Trip {
  id: string
  company_id: string
  vehicle_id: string
  driver_id: string
  trailer_id: string | null
  client_id: string | null
  origin: string
  destination: string
  distance_km: number | null
  cargo_type: string
  is_round_trip: boolean
  status: TripStatus
  start_odometer_km: number | null
  end_odometer_km: number | null
  freight_cost: number | null
  advance_payment: number
  started_at: string | null
  ended_at: string | null
  custom_data?: Record<string, unknown>
  created_at: string
  updated_at: string
}

export interface TripExpense {
  id: string
  trip_id: string
  concept_id: string
  amount: number
  notes: string | null
  recorded_by: string | null
  created_at: string
}

export interface TripPayroll {
  id: string
  trip_id: string
  base_salary: number
  meal_allowance: number
  holiday_bonus: number
  return_bonus: number
  bonuses: number
  advance_payment: number
  total_to_pay: number
  created_at: string
  updated_at: string
}

export interface TripWithDetails extends Trip {
  expenses: TripExpense[]
  payroll: TripPayroll | null
}

export interface TripClosePreview {
  trip_days: number
  base_salary: number
  meal_allowance: number
  holiday_bonus: number
  holiday_count: number
  return_bonus: number
  bonuses: number
  advance_payment: number
  total_to_pay: number
  freight_cost: number | null
  missing_pay_rate: boolean
}
